crates/storage/src/lib.rs

```diff
@@ -183,11 +181,121 @@ impl Storage {
     }
 
+    /// Performs the database schema migration and returns a [storage manager](StorageManager).
+    ///
+    /// This should be called __once__ at the start of the application,
+    /// and passed to the various components which require access to the database.
+    pub fn migrate(self) -> anyhow::Result<StorageManager> {
+        let mut connection =
+            rusqlite::Connection::open(&self.database_path).context("Opening DB for migration")?;
+
+        // Migration is done with rollback journal mode. Otherwise dropped tables
+        // get copied into the WAL which is prohibitively expensive for large
+        // tables.
+        setup_journal_mode(&mut connection, JournalMode::Rollback)
+            .context("Setting journal mode to rollback")?;
+        setup_connection(&mut connection, JournalMode::Rollback)
+            .context("Setting up database connection")?;
+
+        migrate_database(&mut connection).context("Migrate database")?;
+
+        // Set the journal mode to the desired value.
+        setup_journal_mode(&mut connection, self.journal_mode).context("Setting journal mode")?;
+
+        // Validate that configuration matches database flags & default to the DB setting.
+        let prune_merkle_tries = self.determine_if_pruning_is_enabled(&mut connection)?;
+        if prune_merkle_tries {
+            tracing::info!("Merkle trie pruning enabled");
+        } else {
+            tracing::info!("Merkle trie pruning disabled");
+        }
+
+        connection
+            .close()
+            .map_err(|(_connection, error)| error)
+            .context("Closing DB after migration")?;
+
+        Ok(StorageManager {
+            database_path: self.database_path,
+            journal_mode: self.journal_mode,
+            bloom_filter_cache: Arc::new(bloom::Cache::with_size(self.bloom_filter_cache_size)),
+            prune_merkle_tries,
+        })
+    }
+
+    /// Returns if trie pruning should be enabled.
+    ///
+    /// Takes the configured value (if present) as an input.
+    /// - If there is no explicitly requested configuration (default): uses the DB setting.
+    /// - If there's an explicitly requested setting: uses it if matches DB setting, enables
+    ///   pruning and sets flag only if tries are empty.
+    fn determine_if_pruning_is_enabled(
+        &self,
+        connection: &mut rusqlite::Connection,
+    ) -> anyhow::Result<bool> {
+        let prune_flag_is_set = connection
+            .query_row(
+                "SELECT 1 FROM storage_flags WHERE flag = 'prune_tries'",
+                [],
+                |_| Ok(()),
+            )
+            .optional()
+            .map(|x| x.is_some())?;
+
+        let tries_are_non_empty: bool = connection.query_row(
+            "SELECT EXISTS (SELECT 1 FROM trie_storage) OR EXISTS (SELECT 1 FROM trie_contracts) OR EXISTS (SELECT 1 FROM trie_class)",
+            [],
+            |row| row.get(0),
+        )?;
```

Comment:
This works; but I'm thinking about the ergonomics long term here.

Ideally I think we would:

1. Set the flag the very first time the db is created
2. Only check the flag matches the provided one

Currently we can't do this because the rusqlite `Connection::open` by default enables the `SQLITE_OPEN_CREATE` flag.

What do you think about changing the first migration step to:

```rust
let mut connection = {
    // Try open existing db first
    let mut open_flags = rusqlite::OpenFlags::default();
    open_flags.remove(SQLITE_OPEN_CREATE);

    // TODO: actually ensure that the error is file not found..
    if let Ok(connection) = Connection::open_with_flags(open_flags) {
        break connection;
    }
    
    // Database file should be created, and flags should be set.
    let mut connection = Connection::open()?;
    // TODO: set the flag(s).
};
```

Reply:
I've pushed something similar as an update: https://github.com/eqlabs/pathfinder/pull/1747/commits/9ba6f1d5f5dd1c8bd94a9d555fdb05488113521e

Does that make more sense to you?

Reply:
Its better; I don't like that the verification does both verification but also sometimes writes the flags.

```
// This also sets the one-time flags when creating the database.
let connection = open_or_create(flags)?;[
...
migratate(connection)?;

// Only verification/ensure db makes sense with options.
verify_config(connection)?;
```

But now that I consider the issue more closely - its that you need the migrations to be completed before you can write the flags? So one would have to bundle the open/create and migration into one step.. bleg.

Okay 👍 I give up 😬

Reply:
Looks like this breaks enabling pruning for the in-memory DB created by `StorageBuilder::in_memory()`. That's because we do have to pre-create the in-memory DB and keep a connection open to that while `migrate()` is running. That's something we can take care of after merging this PR though?